bot/utils.py

```diff
@@ -45,3 +57,100 @@ def _convert_keys(self):
         for k in list(self.keys()):
             v = super(CaseInsensitiveDict, self).pop(k)
             self.__setitem__(k, v)
+
+
+async def paginate(lines: Iterable[str], ctx: Context, embed: Embed,
+                   prefix: str = "", suffix: str = "", max_size: int = 500, empty: bool = True,
+                   restrict_to_user: User = None):
+    """
+    Use a paginator and set of reactions to provide pagination over a set of lines. The reactions are used to
+    switch page, or to finish with pagination.
+
+    When used, this will send a message using `ctx.send()` and apply a set of reactions to it. These reactions may
+    be used to change page, or to remove pagination from the message. Pagination will also be removed automatically
+    if no reaction is added for five minutes (300 seconds).
+
+    >>> embed = Embed()
+    >>> embed.set_author(name="Some Operation", url=url, icon_url=icon)
+    >>> await paginate(
+    ...     (line for line in lines),
+    ...     ctx, embed
+    ... )
+
+    :param lines: The lines to be paginated
+    :param ctx: Current context object
+    :param embed: A pre-configured embed to be used as a template for each page
+    :param prefix: Text to place before each page
+    :param suffix: Text to place after each page
+    :param max_size: The maximum number of characters on each page
+    :param empty: Whether to place an empty line between each given line
+    :param restrict_to_user: A user to lock pagination operations to for this message, if supplied
+    """
+
+    paginator = Paginator(prefix=prefix, suffix=suffix, max_size=max_size)
+    current_page = 0
+
+    for line in lines:
+        paginator.add_line(line, empty=empty)
+
+    embed.description = paginator.pages[current_page]
+
+    message = await ctx.send(embed=embed)
+
+    if len(paginator.pages) <= 1:
+        return  # There's only one page
+
+    embed.set_footer(text=f"Page {current_page + 1}/{len(paginator.pages)}")
+
+    def event_check(reaction_: Reaction, user_: Member):
```

Comment:
I'm not a fan of this function def being in the middle of the other function body. I'd prefer it right at the start..